sonyc_kalman/data.py

```diff
@@ -70,6 +70,105 @@ def load_openl3_time_series(hdf5_path, delta_mins=15, aggr_func=None):
+        test_mask: np.array of int's
+            an array of 1's and 0's that can be used to mask any series. 1 being masked, 0 being selected for the test split
+    '''
+    if mode == 'chronological' or mode == 'c':
+        chrono_mode = True
+    elif mode == 'random_holes' or mode == 'r':
+        chrono_mode = False
+    else:
+        raise ValueError("type must be one of 'random_holes', 'r', 'chronological', or 'c'.")
+    
+    np.random.seed(random_state)
+
+    invalid_idx = np.array(mask, dtype=np.bool)
+    valid_idx = 1 - invalid_idx
+    valid_idx = valid_idx.astype(np.bool)
+    n_total_valid = sum(valid_idx)
+    n_test = int(round(test_ratio * n_total_valid))
+    n_train = n_total_valid - n_test
+    
+    # init by selecting nothing (1) for test and everything to train (0)
+    train_mask = np.zeros_like(mask)
+    test_mask = np.ones_like(mask)
+    if chrono_mode:
+        # first split the data as if all the valid frames are contiguous
+        continuous_split_test = np.array( [1] * n_train + [0] * n_test, dtype=int)
+        # only select regions in valid_idx for test_mask
+        test_mask[valid_idx]  = continuous_split_test
+        # everything else goes in train_mask
+        train_mask = 1 - test_mask
+
+    else:
+        # 'random_holes' mode
+        ## first generate an array that holes the sizes of the holes, with length n_holes
+        hole_sizes = []
+        while sum(hole_sizes) < n_test:
+            rand_size = int(round(np.random.normal(hole_mean, hole_std)))
+            if rand_size >= min_hole_size:
+                hole_sizes.append(rand_size)
+
+        if sum(hole_sizes) > n_test:
+            hole_sizes.pop()
+            last_size = n_test - sum(hole_sizes)
+            hole_sizes.append(last_size)
+
+        ## next, decide where the holes are gonna be, ie, how to split the training data into n_holes + 1 parts
+        n_holes = len(hole_sizes)
+        # pick n_holes positions from all n_train+1 positions: this avoids two adjacent holes
+        hole_start_positions = np.sort(np.random.choice(n_train+1, n_holes, replace=False))
+        inter_hole_intervals = [hole_start_positions[0]] + list(np.diff(hole_start_positions))
+        after_last_hole = n_train - hole_start_positions[-1]
+
+        ## then build the continuous_split arrays according to hole_sizes
+        continuous_split_test = []
+        for hole, bread in zip(hole_sizes, inter_hole_intervals):
+            continuous_split_test += bread * [1]
+            continuous_split_test += hole * [0]
+        continuous_split_test += after_last_hole * [1]
+
+        ## finally put things into places
+        continuous_split_test = np.array(continuous_split_test, dtype=int)
+        # only select regions in valid_idx for test_mask
+        test_mask[valid_idx]  = continuous_split_test
+        # everything else goes in train_mask
+        train_mask = 1 - test_mask
```

Comment:
An issue here (assuming I understand what's going on correctly) is that because we are creating the holes with respect to the valid indices, we risk splitting the holes with respect to the full set of indices. Consider this example with 10 samples
```
|0|1|2|3|4|5|6|7|8|9|
|-|-|-|-|X|-|-|-|-|-|
```
There are 9 valid indices. Say we sample a single hole from valid indices 2-4. The resulting train/valid split would be:

```
|0|1|2|3|4|5|6|7|8|9|
|T|T|V|V|T|V|T|T|T|T|
```

So we've split the validation hole into two, and we also have a training sequence that is a single invalid sample.

We can revisit doing this properly at some point later, but I think right now it's probably easier to just ignore the mask entirely when doing the sampling. Otherwise it becomes difficult trying to sample contiguous valid regions specifically for the validation set.

Reply:
Let me know your thoughts on 0bb2832

Reply:
LGTM